### PR TITLE
Floating action button and Realm DB implemented

### PR DESCRIPTION
**Pull request by nicolbidone**
Screenshot_1554324645

### CODE REVIEW

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt [hunk 1]
```diff
@@ -1,23 +1,37 @@
...
 import com.puzzlebench.clean_marvel_kotlin.presentation.mvp.CharacterPresenter
 import com.puzzlebench.clean_marvel_kotlin.presentation.mvp.CharacterView
 import io.realm.Realm
+import kotlinx.android.synthetic.main.activity_main.*
 
 open class MainActivity : BaseRxActivity() {
 
     val getCharacterServiceUseCase = GetCharacterServiceUseCase(CharacterServicesImpl())
-    val presenter = CharacterPresenter(CharacterView(this), getCharacterServiceUseCase, subscriptions)
+    val presenter = CharacterPresenter(CharacterView(this), CharacterModel(), getCharacterServiceUseCase, subscriptions)
```
Comment (lucianoeli) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt]: getCharacterServiceUseCase shoudnt be passed as a paramiter of the presenter.
It should be a method inside the model.

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt [hunk 2]
```diff
@@ -33,7 +33,7 @@ class CharacterPresenterTest {
 
         getCharacterServiceUseCase = GetCharacterServiceUseCase(characterServiceImp)
         val subscriptions = mock(CompositeDisposable::class.java)
-        characterPresenter = CharacterPresenter(view, getCharacterServiceUseCase, subscriptions)
+        characterPresenter = CharacterPresenter(view, getCharacterServiceUseCase, subscriptions, CharacterModel("text"), )
```
Comment (lucianoeli) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt]: you can remove "subscriptions" on both presenters

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt [hunk 3]
```diff
@@ -5,19 +5,30 @@ import com.puzzlebench.clean_marvel_kotlin.presentation.base.Presenter
 import io.reactivex.android.schedulers.AndroidSchedulers
 import io.reactivex.disposables.CompositeDisposable
 import io.reactivex.schedulers.Schedulers
+import io.realm.Realm
 
-class CharacterPresenter(view: CharacterView, private val getCharacterServiceUseCase: GetCharacterServiceUseCase, val subscriptions: CompositeDisposable) : Presenter<CharacterView>(view) {
+
+class CharacterPresenter(view: CharacterView, val model: CharacterModel, private val getCharacterServiceUseCase:
```
Comment (lucianoeli) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt]: same, the use case shoudnt be here,it should be inside the model. Also you can remove the "subscriptions"

**Review by lucianoeli — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt (hunk 1 above)
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt]: delete the subscriptions.. and put the useCase as a parameter in the model

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+class CharacterModel {
+
+    lateinit var realm: Realm
+
+    fun init() {
+        realm = Realm.getDefaultInstance()
+    }
+
+    fun getCharacters(): List<Character> {
+        if (realm.where<RealmCharacter>().findAll().size > 0) {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: use constants pls

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            return realmGetter()
+        }
+        return mutableListOf()
+    }
+
+    private fun realmGetter(): List<Character> {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: this is a problem of the UseCase.. 
the model call the useCase, and the useCase call the DB

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                    fin.description,
+                    Thumbnail(fin.thumbnail!!.path, fin.thumbnail!!.extension)
+            )
+            lis.add(cha)
+        }
+
+        return lis
+    }
+
+    fun realmSetter(characters: List<Character>) {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: same

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                } else {
+                    showStatus(CONSTS.MESSAGE_EXISTING_OBJECT)
+                }
+            }
+
+        }
+
+    }
+
+    fun close() {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: same

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt (hunk 3 above)
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt]: the getCharacterServiceUseCase must be a model parameter

File: app/src/main/res/layout/activity_main.xml [hunk 8]
```diff
@@ -18,5 +17,13 @@
         android:layout_width="match_parent"
         android:layout_height="match_parent" />
 
+    <android.support.design.widget.FloatingActionButton
+            android:id="@+id/fab"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="end|bottom"
+            android:layout_margin="16dp"
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_main.xml]: use dimens pls

File: app/src/main/res/layout/activity_main.xml [hunk 9]
```diff
@@ -18,5 +17,13 @@
...
+            android:src="@drawable/ic_save_info"
+    />
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_main.xml:27]: this must be in the previous line

File: app/src/main/res/layout/character_cards_layout.xml [hunk 10]
```diff
@@ -29,3 +29,5 @@
     </android.support.v7.widget.CardView>
 
 </RelativeLayout>
+
+
```
Comment (SartiMau) [inline app/src/main/res/layout/character_cards_layout.xml:33]: delete unnecessary line

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt (hunk 2 above)
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt]: getCharacterServiceUseCase must be a model parameter.. and you can delete the subscriptions here.

Why the model get a hardcoded string? delete it pls

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt (hunk 4 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: Done

File: app/src/main/res/layout/activity_main.xml (hunk 8 above)
Comment (nicolbidone) [reply app/src/main/res/layout/activity_main.xml]: Done

File: app/src/main/res/layout/character_cards_layout.xml (hunk 10 above)
Comment (nicolbidone) [reply app/src/main/res/layout/character_cards_layout.xml:33]: Done

File: app/src/main/res/layout/activity_main.xml (hunk 9 above)
Comment (nicolbidone) [reply app/src/main/res/layout/activity_main.xml:27]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt (hunk 2 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt (hunk 1 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt]: Done

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt (hunk 2 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt (hunk 3 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt (hunk 1 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/MainActivity.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt (hunk 5 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt (hunk 6 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt (hunk 7 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterModel.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt (hunk 3 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenter.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+        var lis = mutableListOf<Character>()
+        val findAllAsync = realm.where<RealmCharacter>().findAllAsync()
+        for (fin in findAllAsync) {
+
+            var character = Character(
+                    fin.id,
+                    fin.name,
+                    fin.description,
+                    Thumbnail(fin.thumbnail!!.path, fin.thumbnail!!.extension)
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: try not to use "!!"

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+import com.puzzlebench.clean_marvel_kotlin.REALM_TAG
+import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
+import com.puzzlebench.clean_marvel_kotlin.domain.model.RealmCharacter
+import com.puzzlebench.clean_marvel_kotlin.domain.model.RealmThumbnail
+import com.puzzlebench.clean_marvel_kotlin.domain.model.Thumbnail
+import io.realm.Realm
+import io.realm.kotlin.createObject
+import io.realm.kotlin.where
+
+class CharacterStoredImpl {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: Make an interface for this. The useCase will call the interface and this class will implement it

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt [hunk 13]
```diff
@@ -33,7 +33,7 @@ class CharacterPresenterTest {
 
         getCharacterServiceUseCase = GetCharacterServiceUseCase(characterServiceImp)
         val subscriptions = mock(CompositeDisposable::class.java)
-        characterPresenter = CharacterPresenter(view, getCharacterServiceUseCase, subscriptions, CharacterModel("text"), )
+        characterPresenter = CharacterPresenter(view, getCharacterServiceUseCase, subscriptions, CharacterModel("text", getCharacterStoredUseCase), )
```
Comment (SartiMau) [inline app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt]: Why do you have the "text" parameter hardcoded here??

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt (hunk 11 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+        var lis = mutableListOf<Character>()
+        val findAllAsync = realm.where<RealmCharacter>().findAllAsync()
+        for (fin in findAllAsync) {
+
+            var character = Character(
+                    fin.id,
+                    fin.name,
+                    fin.description,
+                    Thumbnail(fin.thumbnail?.path ?: "", fin.thumbnail?.extension ?: "")
```
Comment (lucianoeli) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: constantsss

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    override fun setCharacters(characters: List<Character>) {
+        val realm = Realm.getDefaultInstance()
+
+        realm.executeTransaction { realmObject ->
+            for (character in characters) {
+                val hasId = realmObject.where<RealmCharacter>().equalTo(ID_TEXT, character.id).findFirst()
+                if (hasId == null) {
+                    val aux = realmObject.createObject<RealmCharacter>(character.id)
```
Comment (lucianoeli) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt:48]: you could try a variable name more descriptive. Like "realmCharacter"

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt [hunk 16]
```diff
@@ -11,16 +12,12 @@ import com.puzzlebench.clean_marvel_kotlin.presentation.extension.showToast
...
 
     private val activityRef = WeakReference(activity)
+    private var adapter = CharacterAdapter { character -> showFragmentDialog(character) }
 
-    var adapter = CharacterAdapter { character -> showFragmentDialog(character) }
-
-    fun init() {
+    override fun init() {
         val activity = activityRef.get()
         if (activity != null) {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt]: use the safe call

File: app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt (hunk 13 above)
Comment (nicolbidone) [reply app/src/test/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterPresenterTest.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt (hunk 12 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt [hunk 17]
```diff
@@ -29,7 +26,7 @@ class CharacterView(activity: MainActivity) {
         }
     }
 
-    fun showToastNoItemToShow() {
+    override fun showToastNoItemToShow() {
         val activity = activityRef.get()
         if (activity != null) {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt]: same

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt [hunk 18]
```diff
@@ -26,7 +27,7 @@ class CharacterStoredImpl {
                     fin.id,
                     fin.name,
                     fin.description,
-                    Thumbnail(fin.thumbnail!!.path, fin.thumbnail!!.extension)
+                    Thumbnail(fin.thumbnail?.path ?: "", fin.thumbnail?.extension ?: "")
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: no hardcode pls

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmCharacter.kt [hunk 19]
```diff
@@ -1,12 +1,13 @@
...
+import com.puzzlebench.clean_marvel_kotlin.ZERO_VALUE
 import io.realm.RealmObject
 import io.realm.annotations.PrimaryKey
 
 open class RealmCharacter(
-        @PrimaryKey var id: Int = CONSTS.ZERO_VALUE,
-        var name: String = CONSTS.EMPTY_VALUE,
-        var description: String = CONSTS.EMPTY_VALUE,
+        @PrimaryKey var id: Int = ZERO_VALUE,
+        var name: String = EMPTY_VALUE,
```
Comment (lucianoeli) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmCharacter.kt:10]: you could use de constatnts file you create

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmThumbnail.kt [hunk 20]
```diff
@@ -1,13 +1,14 @@
...
+import com.puzzlebench.clean_marvel_kotlin.THUMBNAIL_ID
 import io.realm.RealmObject
 import io.realm.RealmResults
 import io.realm.annotations.LinkingObjects
 
 open class RealmThumbnail : RealmObject() {
-    var path: String = CONSTS.EMPTY_VALUE
-    var extension: String = CONSTS.EMPTY_VALUE
-    @LinkingObjects(CONSTS.THUMBNAIL_ID)
+    var path: String = EMPTY_VALUE
```
Comment (lucianoeli) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmThumbnail.kt:10]: same

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/fragment/CharacterFragment.kt [hunk 21]
```diff
@@ -8,25 +8,28 @@ import android.graphics.drawable.ColorDrawable
...
+import com.puzzlebench.clean_marvel_kotlin.fragment.mvp.FragmentModel
+import com.puzzlebench.clean_marvel_kotlin.fragment.mvp.FragmentPresenter
+import com.puzzlebench.clean_marvel_kotlin.fragment.mvp.FragmentView
 import com.puzzlebench.clean_marvel_kotlin.presentation.MainActivity
 import com.puzzlebench.clean_marvel_kotlin.presentation.extension.getImageByUrl
 import kotlinx.android.synthetic.main.fragment_character.view.*
 
 class CharacterFragment : DialogFragment() {
 
+    private val getSingleCharacterServiceUseCase = GetSingleCharacterServiceUseCase(CharacterServicesImpl(), character.id)
```
Comment (lucianoeli) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/fragment/CharacterFragment.kt:27]: i think you can remove this val

**Review by lucianoeli — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt (hunk 18 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt (hunk 14 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmCharacter.kt (hunk 19 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmCharacter.kt:10]: import com.puzzlebench.clean_marvel_kotlin.EMPTY_VALUE
import com.puzzlebench.clean_marvel_kotlin.ZERO_VALUE
Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmThumbnail.kt (hunk 20 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmThumbnail.kt:10]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/fragment/CharacterFragment.kt (hunk 21 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/fragment/CharacterFragment.kt:27]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmCharacter.kt (hunk 19 above)
Comment (lucianoeli) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/RealmCharacter.kt:10]: my bad

**Review by lucianoeli — APPROVED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt (hunk 16 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt (hunk 17 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt (hunk 15 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterStoredImpl.kt:48]: Done

**Review by SartiMau — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Pull request merged**